Implement Variable.cuda and Variable.type using ATen

colesbury opened the pull request:
```
This adds an optional async flag to Tensor::copy_, which attempts to do
a non-blocking copy if the one of the tensors is in pinned memory and
the other is a CUDA tensor.
```

I've also changed the signature of `s_copy` to match other functions: `self` is the first parameter and it has an `_` suffix to signify that it's in-place.

colesbury committed:
Implement Variable.cuda using ATen

This adds an optional async flag to Tensor::copy_, which attempts to do
a non-blocking copy if the one of the tensors is in pinned memory and
the other is a CUDA tensor.

colesbury committed:
Implement Variable.type via ATen

colesbury committed:
Perform cross-device copy in CopyBackwards

Also call torch.cuda._lazy_init() from Variable.cuda()

apaszke commented on aten/src/ATen/Context.cpp:
```diff
@@ -74,4 +74,15 @@ cudaStream_t Context::getCurrentCUDAStream() const {
+#if AT_CUDA_ENABLED()
+  int device;
+  cudaError_t err = cudaGetDevice(&device);
+  if (err == cudaSuccess) {
```
Returning -1 in case of an error is a bit weird, because it's actually a meaningful value for device idx in PyTorch. Why not throw an error?

colesbury replied:
I want the behavior for no GPUs to match the behavior for compiling without CUDA support. So that means both cases should return the same value or both cases should throw. In general, having these functions be safe to call in a context without CUDA is more useful than throwing exceptions.

apaszke commented on tools/autograd/templates/VariableType.cpp:
```diff
@@ -335,26 +335,26 @@ static bool isFloatingPoint(ScalarType s) {
 }
 
-void VariableType::s_copy(const Tensor & src, Tensor & dst) const {
+Tensor & VariableType::s_copy_(Tensor & self, const Tensor & src, bool async) const {
```
uh are we changing src-dst order again...

colesbury replied:
I'm making it consistent with every other method in Type

apaszke commented on torch/csrc/autograd/functions/tensor.cpp:
```diff
@@ -21,7 +21,11 @@ auto CopyBackwards::apply(const variable_list& grads) -> variable_list {
     AutoGPU autoGPU(src_device);
-    grad_inputs[1] = grad.toType(*src_type);
+    if (grad.is_cuda() && grad.get_device() != src_device) {
+      grad_inputs[1] = src_type->copy(grad);
```
is `type->copy(grad)` equivalent to creating a clone, but of type `type`?

colesbury replied:
Yes

apaszke reviewed (commented):
Haven't reviewed the ATen changes.

ezyang commented:
If we're losing the 'async' identifier in Python, should we also avoid using it in the C++ ATen side?

gchanan commented on tools/autograd/templates/python_variable_methods.cpp:
```diff
@@ -279,34 +280,63 @@ static PyObject * THPVariable_invert(PyObject* self, PyObject* args) {
+  if (!self.type().is_cuda() && type.is_cuda()) {
+    return type.copy(self, async);
+  }
+  return self.toType(type);
```
why can't we just type.copy here as well?

colesbury replied:
We don't want to force a copy if the types and the devices are the same.

gchanan replied:
Sorry, I expressed this poorly.  I meant what you are doing here seems totally natural and should have better ATen support.  Instead you have to go through all these contortions to figure out if you can avoid a copy.

If at::current_device returns -1 then there is no reason that Tensor.get_device can't, which means at a minimum you can collapse the first two ifs by just looking at the current devices and types (you can do this now by having a "safe_device = self.is_cuda() ? self.get_device : -1" or something.

colesbury replied:
I've combined the if statements.

1) I like the idea of making `Tensor.get_device()` return -1 for CPU tensors. I think it'll simplify the calling code in a bunch of places. I'll do that in a subsequent PR
2) It would be nice to push all of this logic down to ATen. The problem is that ATen doesn't have the "auto gpu" semantics (because it might affect perf), so it doesn't translate as well.

gchanan replied:
Sounds good.

gchanan commented on tools/autograd/templates/python_variable_methods.cpp:
```diff
@@ -279,34 +280,63 @@ static PyObject * THPVariable_invert(PyObject* self, PyObject* args) {
   AutoNoGIL no_gil;
-  AutoGPU auto_gpu(self);
-  return self.toBackend(backend);
+  AutoGPU auto_gpu(device != -1 ? device : self.type().is_cuda() ? self.get_device() : -1);
```
this is different behavior than before right?

```
>>> torch.randn(2).cuda(1).cuda(-1)

-0.6868
-0.5133
[torch.cuda.FloatTensor of size 2 (GPU 0)]
```

(i.e. before -1 meant 0, but now it means current device).

colesbury replied:
Good catch -- I'll fix the code. It's a little weird because `tensor.cuda()` defaults to the current device, while every other method on tensor (including `.byte()`, `.float()`, etc.) defaults to the tensor's device.

gchanan commented:
On the 'async' identifier issue that @ezyang brings up above, I think we should change it to match whatever we do in python.  I personally prefer an indication that this _may_ be async (since we don't do async for cpu-to-cpu copies for example).  Something like maybeAsync, but I understand if people think that is too verbose.

colesbury commented on aten/src/ATen/Context.h:
```diff
@@ -86,4 +87,8 @@ static inline bool hasCUDA() {
   return globalContext().hasCUDA();
 }
 
+static inline int64_t current_device() {
```
My thinking is that we want the C++ function names to match their Python equivalents. Since we use `snake_case` in Python, we should prefer `snake_case` in C++.

It would be nice to make the naming completely consistent within ATen, but that seems low priority.

@zdevito if you have any naming preferences.

gchanan replied:
Sure, I don't really have a strong preference as long as they are eventually consistent.

colesbury commented on aten/src/ATen/templates/Functions.h:
```diff
@@ -10,10 +10,8 @@
-  dst.type().copy(src,dst);
-  return dst;
+static inline Tensor & copy_out(Tensor & result, const Tensor & src) {
+  return result.copy_(src);
```
I don't think this function should exist actually. We don't have `_out` variants for in-place functions.

colesbury commented on torch/csrc/utils/tensor_types.h:
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#include <ATen/ATen.h>
+#include <Python.h>
```
Copy-pasta: we don't need it.

RE: ValueErrors: they're C++ exceptions. If you're in a C++ context, what does it matter if the type is ValueError or just std::runtime_error? Thrown exceptions behave the same.

gchanan replied:
maybe ValueError is a bad example, but there's a bunch of python/PyObject* stuff in Exceptions and it's imo nicer to have a conceptually simpler model of "stuff in this folder/name is for handling python things"

colesbury replied:
Yeah, I think we need to re-organize the code. I'd like to wait until after we delete Tensor though.

colesbury committed:
Changes from review:

 - remove copy_out
 - remove unnecessary include
 - fix default device for .cuda()

colesbury committed:
Combine if statements in dispatch_type
